contracts/HubPool.sol

```diff
@@ -370,39 +392,61 @@ contract HubPool is HubPoolInterface, Testable, Lockable, MultiCaller, Ownable {
+     * @notice Stores mapping from origin chain ID + token + destination chain ID to token. This mapping corresponds
+     * to token transfer routes and can be used for both HubPool => SpokePool rebalances and SpokePool ==> SpokePool
+     * deposits. Callable only by owner.
+     * @param originChainId Chain where token transfer originates.
+     * @param destinationChainId Chain where token transfer ends up.
+     * @param originToken Token sent from origin chain.
+     * @param destinationToken Token received at destination chain.
+     * @param depositsEnabled Set to true to whitelist this route for deposits, set to false if caller just wants to
+     * map the origin token + origin ID + destination ID to the destination token address and wants to explicitly
+     * disable this route for deposits.
      */
+
     function whitelistRoute(
         uint256 originChainId,
         uint256 destinationChainId,
         address originToken,
         address destinationToken,
-        bool enableRoute
-    ) public override onlyOwner nonReentrant {
-        if (enableRoute)
-            whitelistedRoutes[_whitelistedRouteKey(originChainId, originToken, destinationChainId)] = destinationToken;
-        else delete whitelistedRoutes[_whitelistedRouteKey(originChainId, originToken, destinationChainId)];
+        bool depositsEnabled
+    ) public override onlyOwner {
+        whitelistedRoutes[_whitelistedRouteKey(originChainId, originToken, destinationChainId)] = Route({
+            destinationToken: destinationToken,
+            depositsEnabled: depositsEnabled
+        });
+        emit WhitelistRoute(originChainId, destinationChainId, originToken, destinationToken, depositsEnabled);
+    }
 
-        // Whitelist the same route on the origin network.
+    /**
+     * @notice Send cross-chain message to SpokePool on originChainId to enable or disable mapping from origin
+     * origin token to destination chain ID. If this path is enabled on the SpokePool, then relays are allowed to be
+     * sent via that path. The destination token address mapped to the origin token can be set by `whitelistRoute` and
+     * fetched by off-chain relayers. The SpokePool doesn't need to know which destination tokens are mapped to origin
+     * tokens as this protocol assumes that relayers use the correct whitelisted route. Callable only by owner,
+     * who has the responsibility to make sure  that calls to `whitelistRoute` are made in concert with these calls,
+     * otherwise the enabled path information will become out of sync between SpokePool and HubPool.
+     * @param originChainId Chain where token transfer originates.
+     * @param destinationChainId Chain where token transfer ends up.
+     * @param originToken Token sent from origin chain.
+     * @param depositsEnabled Set to True/False to enable/disable route on SpokePool for deposits.
+     */
+    function relayWhitelistRoute(
```

Comment:
I'm not sure of a better way to do this, but it does seem a bit clunky and error-prone to have two separate methods for setting the mapping and sending that setting to the spoke since we would _almost always_ want to do these together. No real suggestion here, just thinking that this might add frustration to the management of the protocol as the number of chains and tokens increases.

Reply:
What if we added a third function to combine the two methods? The other way is to add flags as function params to fine tune the logic. I can see how we'd want to only whitelist on the HubPool/SpokePool in certain situations so I think this flexibility is important

Reply:
Technically we could call both simultaneously with multicall, it just might be annoying, especially if this is called via a gnosis multisig or governance system.

I could see a flag param also making sense.

Curious what @chrismaree thinks.

Reply:
The trade off is that the flag param adds gas, but condenses three functions into one. I probably lean towards usability and adding the following bool params to the `whitelistRoute` call: `bool relayToSpokePool, bool setInHub`

Reply:
I had another idea. What if we didn't store the `depositsEnabled` boolean on `HubPool` and only stored it on `SpokePools`? This would force relayers to use SpokePool whitelisted routes as the source of truth for determining which deposit routes are enabled, while the HubPool whitelist would only store pool rebalance routes.

The advantage is that the separation between pool rebalance and deposit routes would be more cleanly defined.

Reply:
I see. The hub pool would still need an address map, right?

Reply:
I think that your proposal for adding a bool to `whitelistRoute` that, when set, also enables the whitelist route makes a lot of sence as I agree having these separate as two functions is a pain as we will almost always be calling both.

>  had another idea. What if we didn't store the depositsEnabled boolean on HubPool and only stored it on SpokePools? This would force relayers to use SpokePool whitelisted routes as the source of truth for determining which deposit routes are enabled, while the HubPool whitelist would only store pool rebalance routes.

I like this idea but I worry that it drastically increases the L1->L1 calls we need to do as well as conflating the information on the L2 side. This might not be as bad as I think though...There are some other nice benefits to this as well. When making a deposit the `FundsDeposited` will then _also_ be able to contain the `DestinationToken` which right now it does not as this information is not there.

your proposal would look something like this, right?
change `setEnableRoute` to have `destinationToken` as well and then modify `enabledDepositRoutes` to be `originToken->destinationChainId->destinationStruct` where `destinationStruct` contains `destinationToken` and `depositsEnabledenabled`. Then, we can remove this bit of logic from the `HubPool` as the only place that needs to know if a route is enabled is the deposit side (the hub pool does not care!). 

In fact in this pattern what is required to be left on the `HubPool`? almost nothing, no?

Reply:
actually I was proposing to store `originToken -> destinationChainId -> depositsEnabled` on the SpokePool (i.e. keep it as is now) and then store `hash(originChainId, originToken, destinationChainId) -> destinationToken` on the HubPool since we need the destination token for pool rebalances.

But I see why you now recommend we also store `destinationToken` on the SpokePool. Thoughts?

Reply:
@chrismaree , @mrice32 and I talked off-line and we decided to remove the `destinationToken` from the SpokePool as we think the information is redundant and could lead to a situation where we set up the pairwise mappings incorrectly between SpokePools. This could cause routes between SpokePools to be inconsistent. Instead, we can use the `poolRebalanceRoute` on the `HubPool` as the source of truth for L1 --> L2 token mappings.